torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -685,170 +697,221 @@ def __init__(
-        and each rank away from highest rank adds 2 warmup steps due to:
-            - one happened before highest rank's warmup started,
-            - one waiting for backward result to trickle down from highest rank
-
-        TODO: Interleaved 1F1B does not support using _sorted_batch_p2p()
-        because it requires recvs and sends from different peers
-        to execute in the same coalesced operation. As a result, this schedule does
-        not support models with skip connections.
-        """
-        arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs, losses)
-
-        # increment warmup_steps by 2 for each hop away
-        warmup_steps = (self.n_local_stages - 1) * self.pp_group_size
-        warmup_steps += 2 * ((self.pp_group_size - 1) - self.rank)
-        warmup_steps = min(warmup_steps, self._n_microbatches * self.n_local_stages)
-        fwd_bwd_steps = (self.n_local_stages * self._n_microbatches) - warmup_steps
-        cooldown_steps = (self.n_local_stages * self._n_microbatches) - fwd_bwd_steps
-
+        self.group = stages[0].group
+
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [Action(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: Dict[int, List[Optional[_Action]]] = {}
+        # ========================================================================
+        for rank in range(self.pp_group_size):
+            rank_ops = self._calculate_single_rank_operations(rank)
+            self.pipeline_order[rank] = rank_ops
+
+    def _calculate_single_rank_operations(self, rank) -> List[Optional[_Action]]:
+        def get_rank_warmup_ops(rank):
+            # Warms up operations for last stage
+            warmups_ops_last_stage = (self.n_local_stages - 1) * self.pp_group_size
+            # Increment warmup operations by 2 for each hop away from the last stage
+            warmup_ops = warmups_ops_last_stage + 2 * ((self.pp_group_size - 1) - rank)
+            # We cannot have more warmup operations than there are number of microbatches, so cap it there
+            return min(warmup_ops, self._n_microbatches * self.n_local_stages)
+
+        warmup_ops = get_rank_warmup_ops(rank)
+        # fwd_bwd_ops should encompass the remaining forwards
+        fwd_bwd_ops = (self.n_local_stages * self._n_microbatches) - warmup_ops
+        # cooldown_ops should encompass the remaining backwards
+        cooldown_ops = (self.n_local_stages * self._n_microbatches) - fwd_bwd_ops
+
+        # Assert that each stage will go through the entire pipeline of micrbatches
+        # twice, once for forward and once for backward
         assert (
-            warmup_steps + fwd_bwd_steps * 2 + cooldown_steps
+            warmup_ops + fwd_bwd_ops * 2 + cooldown_ops
```

Comment:
you could probably delete all this assertion based on the refactor i suggested above.